### PR TITLE
Story/quad 149

### PR DESCRIPTION
**Pull request by v1kd**
Please review.

### CODE REVIEW

File: Quadriga/src/main/java/edu/asu/spring/quadriga/service/network/impl/NetworkTransformer.java [hunk 1]
```diff
@@ -35,45 +28,60 @@
...
+    private EventParser parser;
 
     @Autowired
-    private EventParser parser;
+    private ElementEventTypeDownloadService elementEventTypeDownloadService;
 
     private static final Logger logger = LoggerFactory
             .getLogger(NetworkTransformer.class);
 
+    private static int NUM_THREADS = 4;
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/service/network/impl/NetworkTransformer.java]: Let's make this configurable. Depending on the capabilities of the machine, admins might want to adjust this.

File: Quadriga/src/main/java/edu/asu/spring/quadriga/service/network/impl/NetworkTransformer.java [hunk 2]
```diff
@@ -35,45 +28,60 @@
...
-                }
-            }
+        if (networkTopNodesList == null || networkTopNodesList.size() == 0) {
+            // return the trasnformed network
+            return transformedNetwork;
         }
 
+        // networkTopNode list has size > 0
+        // Create executors and tasks
+        ExecutorService executorService = Executors.newFixedThreadPool(NUM_THREADS);
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/service/network/impl/NetworkTransformer.java]: Let's use Springs TaskExecutor here and inject it (http://docs.spring.io/spring/docs/current/spring-framework-reference/html/scheduling.html#scheduling-task-executor-usage). If I'm not wrong you create a new thread pool every time the method is called, which kind of goes against the idea of a thread pool.

File: Quadriga/src/main/java/edu/asu/spring/quadriga/service/network/impl/NetworkTransformer.java [hunk 3]
```diff
@@ -35,45 +28,60 @@
...
+                Future<ElementEventsType> future = completionService.take();
+                // take() will block if none are executed
+                ElementEventsType elementEventsType = future.get();
+                parser.parseStatement(networkNodeInfo.getId(), elementEventsType, nodes, links);
+            } catch (InterruptedException | ExecutionException e) {
+                logger.error("Issue while retrieving and marshalling object", e);
+            }
+        }));
+        // finally shutdown the service
+        executorService.shutdown();
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/service/network/impl/NetworkTransformer.java]: Also, let's put all of this code in its own class. The class should have a method that returns a lit of ElementEventsType once all threads are done. Then this method here can call the parser for each element. Does this make sense? This way, the parser does the parsing, your new class does the thread/future handling, and this class just coordinates the transformation. single responsibility principal ;)

File: Quadriga/src/main/java/edu/asu/spring/quadriga/service/network/impl/NetworkTransformer.java (hunk 2 above)
Comment (v1kd) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/service/network/impl/NetworkTransformer.java]: With TaskExecutor I dont think we can use Future or Callable. Is that fine?

**Comment by v1kd** [comment]
Made changes as per your comments.

File: Quadriga/src/main/java/edu/asu/spring/quadriga/service/network/impl/NetworkTransformer.java (hunk 2 above)
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/service/network/impl/NetworkTransformer.java]: i'm pretty sure you could have used  your previous code, as the taskExecutor implemented the same interface as the executorservice. But looks fine too.

**Comment by jdamerow** [comment]
ok to test

**Pull request merged**